kong/plugins/grpc-web/deco.lua

```diff
@@ -0,0 +1,212 @@
+  ["application/grpc-web-text"] = "base64",
+  ["application/grpc-web+proto"] = "plain",
+  ["application/grpc-web-text+proto"] = "base64",
+  ["application/grpc-web+json"] = "plain",
+  ["application/grpc-web-text+json"] = "base64",
+  ["application/json"] = "plain",
+}
+
+local framing_form_mime = {
+  ["application/grpc-web"] = "grpc",
+  ["application/grpc-web-text"] = "grpc",
+  ["application/grpc-web+proto"] = "grpc",
+  ["application/grpc-web-text+proto"] = "grpc",
+  ["application/grpc-web+json"] = "grpc",
+  ["application/grpc-web-text+json"] = "grpc",
+  ["application/json"] = "none",
+}
+
+local msg_encodign_from_mime = {
+  ["application/grpc-web"] = "proto",
+  ["application/grpc-web-text"] = "proto",
+  ["application/grpc-web+proto"] = "proto",
+  ["application/grpc-web-text+proto"] = "proto",
+  ["application/grpc-web+json"] = "json",
+  ["application/grpc-web-text+json"] = "json",
+  ["application/json"] = "json",
+}
+
+
+
+-- parse, compile and load .proto file
+-- returns the parsed info as a table
+local _proto_info = {}
+local function get_proto_info(fname)
+  local info = _proto_info[fname]
+  if info then
+    return info
+  end
+
+  local p = protoc.new()
+  local parsed = p:parsefile(fname)
+
+  info = {}
+
+  for _, srvc in ipairs(parsed.service) do
+    for _, mthd in ipairs(srvc.method) do
+      info[("/%s.%s/%s"):format(parsed.package, srvc.name, mthd.name)] = {
+        mthd.input_type, mthd.output_type
+      }
+    end
+  end
+
+  _proto_info[fname] = info
+
+  p:loadfile(fname)
+  return info
+end
+
+-- return input and output names of the method specified by the url path
+-- TODO: memoize
+local function rpc_types(path, protofile)
+  if not protofile then
+    return nil
+  end
+
+  local info = get_proto_info(protofile)
+  local types = info[path]
+  if not types then
+    return nil, ("Unkown path %q"):format(path)
+  end
+
+  return types[1], types[2]
+end
+
+
+function deco.new(mimetype, path, protofile)
+  local text_encoding = text_encoding_from_mime[mimetype]
+  local framing = framing_form_mime[mimetype]
+  local msg_encoding = msg_encodign_from_mime[mimetype]
+
+  local input_type, output_type
+  if msg_encoding ~= "proto" then
+    if not protofile then
+      return nil, "transcoding requests require a .proto file defining the service"
+    end
+    input_type, output_type = rpc_types(path, protofile)
+    if input_type == nil and output_type ~= nil then
+      return nil, output_type
+    end
+  end
+
+  return setmetatable({
+    mimetype = mimetype,
+    text_encoding = text_encoding,
+    framing = framing,
+    msg_encoding = msg_encoding,
+    input_type = input_type,
+    output_type = output_type,
+  }, deco)
+end
+
+
+local function frame(ftype, msg)
+  local f_hdr = ffi.new(frame_hdr, ftype, bit.bswap(#msg))
+  return ffi.string(f_hdr, ffi.sizeof(f_hdr)) .. msg
+end
+
+local function unframe(body)
+  if not body then
+    return
+  end
+
+  local hdr = ffi.cast(frame_hdr_p, body)
```

Comment:
What if `body` is too small? Will we have buffer overruns in this case?

Since the job `frame_hdr` achieves is so simple, I would strongly recommend we get rid of the unsafe FFI calls and use the built-in Lua functions such as `string.byte` and logical shift to do the header parsing. It makes us much more resilient against memory errors. See: https://github.com/openresty/lua-resty-websocket/blob/master/lib/resty/websocket/protocol.lua#L84 as an example

Reply:
fixed for `#body` <= sizeof(header), thanks

i did the bit-shifting+`string.char()` dance on the test scripts, since i don't want to use the same looking code to test; as you can see it's far more error-prone.  non-variable headers are the best case use for C structures, IMO.

Reply:
I think it's okay if we add some of the common operations (e.g. converting a 4-byte string to a Lua integer) to the `kong/tools/utils.lua` file. That would allow the test and plugin code to share the same helpers.

It is very uncommon for plugin codes having to use FFI operations directly. For the precise reason that has already been shown above. It is inherently more error-prone and buggy FFI call could cause far more damage than buggy Lua code. In Kong we limit the use of those unless in absolutely performance critical path such as the request router.

Since the bit operations in here is relatively simple and rewriting it using Lua wouldn't even generate much more code than the FFI version, it is a worthy tradeoff.

Reply:
I agree on most points, but definitely not on "It [FFI] is inherently more error-prone".  Quite the contrary in many cases, particularly on protocols.

But we could discuss all day long and get trapped in dogmas.  So, what do you think on using lpack?  it's _very_ useful for protocols, well-tested, and standard on Lua 5.3

Reply:
My argument is that FFI is more error-prone on the memory safety side, for the protocol parsing it may be better, but any code that goes into the main Kong repository has to be considered for both of these aspects.

`lpack` seems nice, first glance it does seems to guard against buffer overruns, which was one of the main concerns for protocol parsers. I actually just found out we already have:

https://github.com/Kong/kong/blob/master/kong-2.0.1-0.rockspec#L30
https://github.com/Kong/lua-pack

which seems to solve this problem perfectly :)

Reply:
done.  it _is_ nicer and shorter.  a pity that the included lua_pack is an old and slightly incompatible fork of lpack, but the essence is the same.

Reply:
@javierguerragiraldez We can update the `lua_pack` library as well since we control the fork. Feel free to submit a PR to the repo.